Reduce usage of locks by TPluginHandler

bendavid opened the pull request:
When running RDF/TTreeProcessorMT with large numbers of threads and files with files on xrootd significant lock contention is still present through ```TPluginHandler::LoadPlugin()``` and ```TPluginHandler::ExecPlugin()``` which are called from ```TFile::Open()```.  This PR minimizes the use of locks, which are now only needed the first time a given TPluginHandler is called for various initialization steps.

This change now completely eliminates the use of the global write lock during RDF event loops also for files read from xrootd and finishes addressing https://github.com/root-project/root/issues/7710

For a simple test case:

Create the files:
```cpp
#include "TFile.h"
#include "TTree.h"
#include "TString.h"
#include <thread>

void testwrite() {

  const unsigned int nfiles = 4000;
  const unsigned int nentries = 1000*1000;

  float outval = 1.;

  for (unsigned int ifile = 0; ifile < nfiles; ++ifile) {
    TFile *fout = TFile::Open(TString::Format("test_%i.root", ifile), "RECREATE");
    TTree *tree = new TTree("tree", "");
    tree->Branch("outval", &outval);
    for (unsigned int ientry = 0; ientry < nentries; ++ientry) {
      tree->Fill();
    }
    tree->Write();
    fout->Close();
  }

}
```

run the event loop:
```python
import ROOT
ROOT.ROOT.EnableImplicitMT()


chain = ROOT.TChain("tree")
chain.Add("root://eoscms.cern.ch//store/cmst3/group/wmass/bendavid/iotestinput/test_*.root")

d = ROOT.ROOT.RDataFrame(chain)
res = d.Sum("outval")

resval = res.GetValue()
print(resval)
```

Testing with 256 threads:

before:
```
        Percent of CPU this job got: 1077%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:51.82
```

after:
```
        Percent of CPU this job got: 5626%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:54.89
```
(in fact there is still some possible lock contention/parameter tuning within the xrootd client itself which could improve things further)

pcanal commented on core/metacling/src/TClingCallFunc.cxx:
```diff
@@ -1998,11 +1998,15 @@ TInterpreter::CallFuncIFacePtr_t TClingCallFunc::IFacePtr()
-         fWrapper = make_wrapper();
+
+      // check if another thread already did it
+      if (!fWrapper) {
```
With this pattern (same with the previous one :) ), `fWrapper` actually should be an atomic pointer.  I.e. another thread might be executing line 1997 as the same time as this thread execute line 2006 (or 2008) and might get a partial value.   Could you update `fWrapper` to become an atomic?

bendavid commented on core/base/inc/TPluginManager.h:
```diff
@@ -146,18 +150,14 @@ friend class TPluginManager;
-
       Longptr_t ret;
-      fCallEnv->Execute(ret);
+      const void *args[nargs] = {&params...};
```
Yes that would explain the test failures.  Let me find some middle ground here...

pcanal replied:
I was probably too terse :(.  It is not reintroducing literally the old behavior but in only a similar fashion.   In this new code, the call will work only in the case where the arguments are const char* or const char* containing text to be parsed to get the real values.  (i.e. matching exactly the argument type doesn't help the new call).  Sorry about the confusion.     

However, if we know no conversion needs to be used, we should be able to use a short cut and use directly the function pointer returned by `TMethodCall::GetCallFunc`.

For the other cases, since `TClingCallFunc` caches the `cling::Value` corresponding to the actual input (and later applying the conversion as needed), the easiest is likely to create a copy of the `TMethodCall` (on the stack?) and use it instead of the 'global' one. 

And this until a stateless code path is added (i.e we could merge the `SetParams` and `Execute` in TMethodCall and thus `CallFunc_SetArguments` and `Execute` in TInterpreter so that they act on a "local"(ish?) `llvm::SmallVector<cling::Value, 8> ` rather than the `TClingCallFunc`

Note: the copying of `TMethodCall` is not 'light'.

bendavid replied:
Yah the problem is that there also plenty of global locks within ```TClingCallFunc``` while doing the conversion etc.  Indeed the "alternative" solution I've implemented for the moment is along the lines of what you suggest with a fast path when no conversion is needed (the verification of this is a bit kludgy though)

bendavid commented:
Solution to take "fast path" only when types can be guaranteed to match is a bit ugly, but something more general would require more comprehensive changes to Cling classes, etc.

bendavid commented:
One remaining issue is the use of the gPluginManagerMutex lock in ```TPluginManager::FindHandler```

A shared lock could solve this.  Does it need to be recursive?  Why is gPluginManagerMutex needed also in ```TPluginHandler::CheckForExecPlugin```?

pcanal commented:
`gPluginManagerMutex` is protecting (mainly but not only) the (shared) container `fHandlers`.  As used, it is recursive because we can not tell what we be the side effects of `SetupCallEnv` (if I remember correctly, it calls `TClass::GetClass` which might load a shared library which might run arbitrary code including something using the `PluginManager`.

> A shared lock could solve this. 

I am not sure what you mean (i.e a straight replacement would not have 'large' performance improvement).

One of the thing that we might be able to do is to set the `fHandlers's TList` to be thread safe (by calling `UseRWLock` on it).  If we do so, the locking will be move from global to down to the internal operation of the `TList`.

bendavid commented:
I mean that using a read/write lock would largely solve the problem, since we could then take only the read lock in ```TPluginManager::FindHandler```

bendavid commented:
(but ok maybe using the TList RWLock as you say would already take care of this)

bendavid commented:
Finally coming back to this one:
-Simplified a little the implementation of ExecPlugin to attempt the "fast path" automatically, while also allowing to explicitly provide the types
-replaced gPluginManagerMutex with the internal RWLock of the underlying TList as suggested

bendavid commented:
Failure for ubuntu20 build looks possibly unrelated to this PR.

bendavid commented:
@pcanal any remaining issues with this one?

bendavid commented:
Rebased to hopefully avoid the previously remaining test failure.

bendavid commented:
I THINK that the remaining failing tests are all unrelated to this PR...

pcanal commented on core/base/inc/TPluginManager.h:
```diff
@@ -138,38 +143,79 @@ friend class TPluginManager;
                  nargs, (unsigned long)sizeof...(params) );
       }
-      return ExecPluginImpl(params...);
+      return ExecPluginImpl<T...>(params...);
```
Out of curiosity, is the `<T...>` necessary or just for clarity (and the surface it looks 'redundant')?

bendavid replied:
Yes, this preserves the type qualifiers and reference types of the parameter types which are part of the comparison later.

Consider the following case:

```cpp
std::vector<double> v;
pluginmanager->ExecPlugin(v);
```

The template argument deduction will lead to ```ExecPlugin<std::vector<double>>(int, const std::vector<double>&)```
Then we'll get ```ExecPluginImpl<std::vector<double>(const std::vector<double>&)```
and finally ```const TClass *tupleclass = TClass::GetClass<std::tuple<std::vector<double>>>();```

However we might know that for the downstream comparison to match we need instead
```const TClass *tupleclass = TClass::GetClass<std::tuple<const std::vector<double>&>>();```

We can achieve this by calling instead 
```cpp
std::vector<double> v;
pluginmanager->ExecPlugin<const std::vector<double>&>(v);
```

Without the extra <T...> the deduction would give ```ExecPluginImpl<std::vector<double>>(const std::vector<double>&)``` instead of ````ExecPluginImpl<const std::vector<double>&>(const std::vector<double>&)``` as we need.

pcanal replied:
Fair enough.  Thanks.

pcanal commented on core/base/inc/TPluginManager.h:
```diff
@@ -138,38 +143,79 @@ friend class TPluginManager;
+      Longptr_t ret;
+
+      // check if types match such that function can be called directly
+      const TClass *tupleclass = TClass::GetClass<std::tuple<T...>>();
```
Strike that :).  It is being compared to another `TClass` (pointer) that is generated at run-time, so this is the only way.

pcanal commented:
This looks great.  Thank you very much.
